### ISSUE TITLE
zlib error

### ISSUE THREAD

**Issue by tmedford** [elastic/apm-agent-ruby#356]
Ever since we installed elastic APM we have noticed these error related to zlib. Any ideas?
![image](https://user-images.githubusercontent.com/4482381/54796472-bde6d280-4c26-11e9-98ca-cd5ccfe1e1bd.png)

**Comment by mikker** [comment]
These could mean the agent has trouble connecting to APM Server. Are you seeing these in production?

**Comment by tmedford** [comment]
Yes. I would question connection errors because we also see this whenever we cycle kubernetes pods out. 
![image](https://user-images.githubusercontent.com/4482381/55447740-f4abd980-5592-11e9-98e3-96f7f6bb8fe9.png)

**Comment by mikker** [comment]
This should be fixed in master. Can you try running with that and let me know if it is?

**Comment by mikker** [comment]
Please reopen if your issue wasn't fixed.

**Comment by slhck** [comment]
I am running into the same issue ever since using the Gem (version 3.2.0).

```
zlib(finalizer): Zlib::GzipWriter object must be closed explicitly.
zlib(finalizer): the stream was freed prematurely.
```

starts occurring in my test suite. Here are some logs from the APM server around that time:

```
apm-server_1     | 2019-12-04T13:29:52.721Z ERROR   [request]   middleware/log_middleware.go:74 internal error  {"request_id": "bdfb41db-c173-4236-b718-259d91829374", "method": "POST", "URL": "/intake/v2/events", "content_length": -1, "remote_address": "172.21.0.12", "user-agent": "elastic-apm-ruby/3.2.0 http.rb/4.2.0 ruby/2.6.5", "response_code": 500, "error": "read tcp 172.21.0.11:8200->172.21.0.12:59274: i/o timeout, context canceled"}
```

What could cause such a timeout? This randomly occurs within dozens of lines that are like this:

```
apm-server_1     | 2019-12-04T13:30:02.708Z INFO    [request]   middleware/log_middleware.go:76 request accepted    {"request_id": "0893399c-51f8-4559-9839-65d2fa86bdbe", "method": "POST", "URL": "/intake/v2/events", "content_length": -1, "remote_address": "172.21.0.12", "user-agent": "elastic-apm-ruby/3.2.0 http.rb/4.2.0 ruby/2.6.5", "response_code": 202}
```

Would it make sense to catch any exceptions and close the writer properly before further bubbling it up the stack?

**Comment by estolfo** [comment]
Thanks for reporting this @slhck! We'll have an update soon.

**Comment by estolfo** [comment]
And just to confirm, you've tried version `3.1.0` of the agent and you don't see the error there?

**Comment by slhck** [comment]
No, I actually haven't, I started from scratch with 3.2.0. Sorry if my post was ambiguous. I haven't used APM before.

**Comment by estolfo** [comment]
I see, thanks for the info. If you're just testing and it's not too much trouble, could you try version `3.1.0` to see if you get the error there? That would help us a lot.

**Comment by slhck** [comment]
I am not getting these errors when running the test suite with the Gem version 3.1.0. These errors happen intermittently though, so I'm not sure if I can reliably reproduce them.

**Comment by mikker** [comment]
This could be coming from the recent changes I did to the `Connection::Http` class to share between Intake and CentralConfig.

**Comment by slhck** [comment]
Is there any way I could reproducibly (from the server side) cause such errors to test it? I'm afraid it's very randomly occurring.
